### PR TITLE
boot: reseal given keys when the respective boot chain has changed

### PR DESCRIPTION
**Pull request by bboozzoo**
It is possible that only the recovery boot chain will be changed when we add a
tried recovery system to the set of tested recovery systems. In this case, the
reseal of fallback objects should still be carried out, even if the run boot
chain has not changed.

### CODE REVIEW

File: boot/seal_test.go [hunk 1]
```diff
@@ -526,8 +567,8 @@ func (s *sealSuite) TestResealKeyToModeenvWithSystemFallback(c *C) {
 		// verify the boot chains data file
 		pbc, cnt, err := boot.ReadBootChains(filepath.Join(dirs.SnapFDEDir, "boot-chains"))
 		c.Assert(err, IsNil)
-		if tc.prevPbc {
-			c.Assert(cnt, Equals, 10)
+		if tc.reuseRunPbc {
+			c.Assert(cnt, Equals, 9)
```
Comment (anonymouse64) [inline boot/seal_test.go:571]: why did this change from 10 -> 9 ?

**Review by anonymouse64 — COMMENTED** [review]
the code change looks fine to me, but I'm confused about one part of the test changes

File: boot/seal_test.go (hunk 1 above)
Comment (bboozzoo) [reply boot/seal_test.go:571]: AFAIU the test was buggy here. In the case when sealedKeys == true and prevPbc == true, we'd hit a continue in line 510 in the old code. Right now it's visible that when a run/recovery mode predictable boot chain is reused, the respective count is not bumped.

**Review by mvo5 — APPROVED** [review]
Thanks, looks fine. One suggestion about layout but feel free to ignore.

File: boot/seal.go [hunk 2]
```diff
@@ -415,26 +422,19 @@ func resealKeyToModeenvSecboot(rootdir string, model *asserts.Model, modeenv *Mo
 	}
 	if !needed {
 		logger.Debugf("reseal not necessary")
-		return nil
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply boot/seal.go:418]: That's a very good suggestion actually. Thanks!

File: boot/seal.go [hunk 3]
```diff
@@ -445,21 +445,24 @@ func resealKeyToModeenvSecboot(rootdir string, model *asserts.Model, modeenv *Mo
 	if err != nil {
 		return err
 	}
-	if !needed {
-		logger.Debugf("fallback reseal not necessary")
-		return nil
-	}
+	if needed {
+		rpbcJSON, _ := json.Marshal(rpbc)
+		logger.Debugf("resealing (%d) to recovery boot chains: %s", nextCount, rpbcJSON)
```
Comment (pedronis) [inline boot/seal.go:450]: should this use nextFallbackCount ? do we have this problem elsewhere here?

File: boot/seal.go [hunk 4]
```diff
@@ -406,35 +406,35 @@ func resealKeyToModeenvSecboot(rootdir string, model *asserts.Model, modeenv *Mo
 		return fmt.Errorf("cannot compose run mode boot chains: %v", err)
 	}
```
Comment (pedronis) [inline boot/seal.go:408]: we need a TODO in  recoveryBootChainsForSystems to fix how it's using model, we probably need to pass more than one model in the testing case

File: boot/seal.go [hunk 5]
```diff
@@ -406,35 +406,35 @@ func resealKeyToModeenvSecboot(rootdir string, model *asserts.Model, modeenv *Mo
...
 
-	// reseal the run object
-	pbc := toPredictableBootChains(append(runModeBootChains, recoveryBootChainsForRunKey...))
-
-	needed, nextCount, err := isResealNeeded(pbc, bootChainsFileUnder(rootdir), expectReseal)
```
Comment (pedronis) [inline boot/seal.go:412]: I just noticed we have typo in the doc comment of isResealNeeded:  s/rease count/reseal count/

**Review by pedronis — APPROVED** [review]
lgtm but we should add a TODO that we need to address soon, also noticed couple of preexisting small issues with doc comments/logs

File: boot/seal.go (hunk 3 above)
Comment (bboozzoo) [reply boot/seal.go:450]: Seems to be the only place.

**Pull request merged**